hbase-migration-tools/bigtable-hbase-replication/README.md

```diff
@@ -0,0 +1,290 @@
+   changes from before the replication was enabled in Cloud Bigtable. This
+   behavior is consistent with enabling replication with an HBase cluster.
+8. Use your operational playbooks to monitor replication metrics. CBT HBase
+   replication library will emit standard HBase replication peer metrics.
+9. Users can also monitor replication status by running `status 'replication'`
+   in HBase shell. The metrics for CBT replication will be under the “peer_id”
+   used in the previous step.
+
+## Error handling
+
+HBase has push based replication. Each region server reads the WAL entries and
+passes them to each replication endpoint. If the replication endpoint fails to
+apply WAL logs, the WAL will accumulate on HBase regions servers.
+
+If a Bigtable cluster is temporarily unavailable, the WAL logs will accumulate
+on region servers. Once the cluster becomes available again, replication can
+continue.
+
+For any non-retryable error, like non-existent column-family, replication will
+pause and WAL logs will build-up. Users should monitor & alert on replication
+progress
+via [HBase replication monitoring](https://hbase.apache.org/book.html#_monitoring_replication_status)
+. The replication library can not skip a replication entry as a single WAL entry
+represents an atomic transaction. Skipping a message will result in divergence
+between source and target tables.
+
+### Incompatible Mutations
+
+Certain HBase delete APIs
+are [not supported on CBT](https://cloud.google.com/bigtable/docs/hbase-differences#mutations_and_deletions)
+. If such mutations are issued on HBase, the CBT client in the replication
+library will fail to propagate them and all replication to CBT endpoint will
+stall. To avoid such stalling, we will log such mutations and skip them.
+Following is a summary of unsupported operations and some supported operations
+that can be modified during WAL write.
+
+|Type of mutation |HBase WAL Write behavior |CBT replication library action|
+|-----------------|-------------------------|------------------------------|
+|DeleteLatestVersion|Resolves the latest version and writes a deletecell with timestamp|Supported, as its normal deletecell
+|DeleteFamilyAtVersion|Not modified|Logged and skipped|
+|DeleteFamilyBeforeTimestamp|Not modified|Converts it to DeleteFamily if timestamp within a configurable threshold.|
+|DeleteFamily|Converts to DeleteFamilyBeforeTimestamp with timestamp=now|See DeleteFamilyBeforeTimestamp|
+|DeleteRow|Converts to DeleteFamilyBeforeTimestamp with timestamp=now for all families|See DeleteFamilyBeforeTimestamp|
+
+The goal of this replication library is to allow migration from HBase to CBT.
+Since CBT will not support these mutations after the users migrate to CBT, they
+are recommended to come up with alternative ways to handle these incompatible
+APIs and not issue them while replication is on.
+
+Another special case is mutations with reverse timestamps. HBase uses a `long`
```

Comment:
link reverse timestamps to public docs https://hbase.apache.org/book.html#reverse.timestamp